### ISSUE TITLE
Speedup GRU by applying the reset gate afterwards?

### ISSUE THREAD

**Issue by gw0** [fchollet/keras#3701]
Would it make sense to modify the GRU layer to apply the reset gate after the recurrent multiply contributes to get a ~10% speedup without loss in accuracy ([analyzed by Baidu](https://svail.github.io/diff_graphs/))? "While this changes the function being computed, it has similar functionality, and produces extremely similar training / accuracy performance in many of our experiments. It also makes all the recurrent multiplies independent of each other."

If I understand correctly, they are suggesting the change at [line L562](https://github.com/fchollet/keras/blob/master/keras/layers/recurrent.py#L562) of:

``` diff
-            hh = self.activation(x_h + K.dot(r * h_tm1 * B_U[2], self.U_h))
+            hh = self.activation(x_h + r * K.dot(h_tm1 * B_U[2], self.U_h))
```

This then enables to group and optimize GEMM computations as recommended and will hopefully produce speedups up to ~40%.

**Comment by carlthome** [comment]
Cool stuff. Do a pr and benchmark it?

**Comment by deontologician** [comment]
Also seems like it should be an option to GRU since it's technically not the same function

**Comment by fchollet** [comment]
Agreed, better to make it optional since we need model backward
compatibility. A benchmark would be welcome.

**Comment by gw0** [comment]
Simply changing the order of reset gate (no other optimizations) does not seem to make any difference.  Unfortunately I do not have any more time to check if other recommended optimizations from Baidu have an effect. Benchmarks are based on `examples/lstm_benchmark.py` with a modified GRU.

```diff
diff --git a/keras/layers/recurrent.py b/keras/layers/recurrent.py
index c593ba0..f8557d4 100644
--- a/keras/layers/recurrent.py
+++ b/keras/layers/recurrent.py
@@ -415,7 +415,7 @@ class GRU(Recurrent):
                  init='glorot_uniform', inner_init='orthogonal',
                  activation='tanh', inner_activation='hard_sigmoid',
                  W_regularizer=None, U_regularizer=None, b_regularizer=None,
-                 dropout_W=0., dropout_U=0., **kwargs):
+                 dropout_W=0., dropout_U=0., reset_afterwards=False, **kwargs):
         self.output_dim = output_dim
         self.init = initializations.get(init)
         self.inner_init = initializations.get(inner_init)
@@ -424,6 +424,7 @@ class GRU(Recurrent):
         self.W_regularizer = regularizers.get(W_regularizer)
         self.U_regularizer = regularizers.get(U_regularizer)
         self.b_regularizer = regularizers.get(b_regularizer)
+        self.reset_afterwards = reset_afterwards
         self.dropout_W, self.dropout_U = dropout_W, dropout_U
 
         if self.dropout_W or self.dropout_U:
@@ -543,8 +544,12 @@ class GRU(Recurrent):
             r = self.inner_activation(x_r + inner_r)
 
             x_h = matrix_x[:, 2 * self.output_dim:]
-            inner_h = K.dot(r * h_tm1 * B_U[0], self.U[:, 2 * self.output_dim:])
-            hh = self.activation(x_h + inner_h)
+            if not self.reset_afterwards:  # classic GRU implementation
+                inner_h = K.dot(r * h_tm1 * B_U[0], self.U[:, 2 * self.output_dim:])
+                hh = self.activation(x_h + inner_h)
+            else:  # reordered Baidu GRU implementation
+                inner_h = r * K.dot(h_tm1 * B_U[0], self.U[:, 2 * self.output_dim:])
+                hh = self.activation(x_h + inner_h)
         else:
             if self.consume_less == 'cpu':
                 x_z = x[:, :self.output_dim]
@@ -559,7 +564,10 @@ class GRU(Recurrent):
             z = self.inner_activation(x_z + K.dot(h_tm1 * B_U[0], self.U_z))
             r = self.inner_activation(x_r + K.dot(h_tm1 * B_U[1], self.U_r))
 
-            hh = self.activation(x_h + K.dot(r * h_tm1 * B_U[2], self.U_h))
+            if not self.reset_afterwards:  # classic GRU implementation
+                hh = self.activation(x_h + K.dot(r * h_tm1 * B_U[2], self.U_h))
+            else:  # reordered Baidu GRU implementation
+                hh = self.activation(x_h + r * K.dot(h_tm1 * B_U[2], self.U_h))
         h = z * h_tm1 + (1 - z) * hh
         return h, [h]
 
@@ -594,7 +602,8 @@ class GRU(Recurrent):
                   'U_regularizer': self.U_regularizer.get_config() if self.U_regularizer else None,
                   'b_regularizer': self.b_regularizer.get_config() if self.b_regularizer else None,
                   'dropout_W': self.dropout_W,
-                  'dropout_U': self.dropout_U}
+                  'dropout_U': self.dropout_U,
+                  'reset_afterwards': self.reset_afterwards}
         base_config = super(GRU, self).get_config()
         return dict(list(base_config.items()) + list(config.items()))
```

```
# Keras 1.0.8+ with TensorFlow 0.9.0 on CPU
Testing mode: consume_less="cpu" + reset_afterwards="False" (classic GRU implementation)
Epoch 1/10
25000/25000 [==============================] - 194s - loss: 0.5096 - acc: 0.7470 - val_loss: 0.3949 - val_acc: 0.8332
Epoch 2/10
25000/25000 [==============================] - 193s - loss: 0.3593 - acc: 0.8467 - val_loss: 0.3715 - val_acc: 0.8416
Epoch 3/10
25000/25000 [==============================] - 194s - loss: 0.2838 - acc: 0.8843 - val_loss: 0.3707 - val_acc: 0.8413
Testing mode: consume_less="cpu" + reset_afterwards="True" (reordered Baidu GRU implementation)
Epoch 1/10
25000/25000 [==============================] - 192s - loss: 0.5180 - acc: 0.7381 - val_loss: 0.3922 - val_acc: 0.8238
Epoch 2/10
25000/25000 [==============================] - 191s - loss: 0.3583 - acc: 0.8460 - val_loss: 0.3914 - val_acc: 0.8313
Epoch 3/10
25000/25000 [==============================] - 191s - loss: 0.2848 - acc: 0.8842 - val_loss: 0.3807 - val_acc: 0.8326

# Keras 1.0.8+ with TensorFlow 0.9.0 on GPU
Testing mode: consume_less="cpu" + reset_afterwards="False" (classic GRU implementation)
Epoch 1/3
25000/25000 [==============================] - 44s - loss: 0.5056 - acc: 0.7494 - val_loss: 0.4093 - val_acc: 0.8262
Epoch 2/3
25000/25000 [==============================] - 44s - loss: 0.3613 - acc: 0.8474 - val_loss: 0.3776 - val_acc: 0.8357
Epoch 3/3
25000/25000 [==============================] - 44s - loss: 0.2857 - acc: 0.8854 - val_loss: 0.3761 - val_acc: 0.8406
Testing mode: consume_less="cpu" + reset_afterwards="True" (reordered Baidu GRU implementation)
Epoch 1/3
25000/25000 [==============================] - 44s - loss: 0.5220 - acc: 0.7340 - val_loss: 0.3924 - val_acc: 0.8294
Epoch 2/3
25000/25000 [==============================] - 44s - loss: 0.3541 - acc: 0.8500 - val_loss: 0.3650 - val_acc: 0.8404
Epoch 3/3
25000/25000 [==============================] - 44s - loss: 0.2829 - acc: 0.8848 - val_loss: 0.3690 - val_acc: 0.8385
Testing mode: consume_less="gpu" + reset_afterwards="False" (classic GRU implementation)
Epoch 1/3
25000/25000 [==============================] - 46s - loss: 0.5415 - acc: 0.7203 - val_loss: 0.3810 - val_acc: 0.8310                
Epoch 2/3
25000/25000 [==============================] - 46s - loss: 0.3602 - acc: 0.8477 - val_loss: 0.4014 - val_acc: 0.8289
Epoch 3/3
25000/25000 [==============================] - 46s - loss: 0.2978 - acc: 0.8746 - val_loss: 0.3910 - val_acc: 0.8243
Testing mode: consume_less="gpu" + reset_afterwards="True" (reordered Baidu GRU implementation)
Epoch 1/3
25000/25000 [==============================] - 46s - loss: 0.5119 - acc: 0.7406 - val_loss: 0.3931 - val_acc: 0.8305
Epoch 2/3
25000/25000 [==============================] - 46s - loss: 0.3582 - acc: 0.8458 - val_loss: 0.3642 - val_acc: 0.8433
Epoch 3/3
25000/25000 [==============================] - 46s - loss: 0.2894 - acc: 0.8822 - val_loss: 0.3889 - val_acc: 0.8323
```

**Comment by nouiz** [comment]
If this give a speed up or not depend of the framework. Does it take this
new opportunity of optimization?

Can you try with Theano? Theano have more automatic uptimization then
tensorflow. So it is possible that this speed up Theano, but not tensorflow.

**Comment by gw0** [comment]
Lets see what does Theano do (same setup)...

```
# Keras 1.0.8+ with Theano 0.8.2 on GPU
Testing mode: consume_less="cpu" + reset_afterwards="False"
Epoch 1/3
25000/25000 [==============================] - 16s - loss: 0.5257 - acc: 0.7298 - val_loss: 0.3742 - val_acc: 0.8328
Epoch 2/3
25000/25000 [==============================] - 16s - loss: 0.3544 - acc: 0.8484 - val_loss: 0.3906 - val_acc: 0.8272
Epoch 3/3
25000/25000 [==============================] - 16s - loss: 0.2849 - acc: 0.8818 - val_loss: 0.3825 - val_acc: 0.8328
Testing mode: consume_less="cpu" + reset_afterwards="True"
Epoch 1/3
25000/25000 [==============================] - 16s - loss: 0.5121 - acc: 0.7411 - val_loss: 0.4201 - val_acc: 0.8012
Epoch 2/3
25000/25000 [==============================] - 16s - loss: 0.3507 - acc: 0.8514 - val_loss: 0.3699 - val_acc: 0.8376
Epoch 3/3
25000/25000 [==============================] - 16s - loss: 0.2967 - acc: 0.8806 - val_loss: 0.3755 - val_acc: 0.8382
Testing mode: consume_less="gpu" + reset_afterwards="False"
Epoch 1/3
25000/25000 [==============================] - 15s - loss: 0.5370 - acc: 0.7156 - val_loss: 0.3813 - val_acc: 0.8286
Epoch 2/3
25000/25000 [==============================] - 15s - loss: 0.3657 - acc: 0.8427 - val_loss: 0.3763 - val_acc: 0.8331
Epoch 3/3
25000/25000 [==============================] - 15s - loss: 0.2828 - acc: 0.8834 - val_loss: 0.3769 - val_acc: 0.8347
Testing mode: consume_less="gpu" + reset_afterwards="True"
Epoch 1/3
25000/25000 [==============================] - 14s - loss: 0.5304 - acc: 0.7254 - val_loss: 0.4040 - val_acc: 0.8239
Epoch 2/3
25000/25000 [==============================] - 14s - loss: 0.3677 - acc: 0.8432 - val_loss: 0.3896 - val_acc: 0.8279
Epoch 3/3
25000/25000 [==============================] - 15s - loss: 0.2849 - acc: 0.8846 - val_loss: 0.3749 - val_acc: 0.8371
```

Although Theano is in general much faster for RNNs, other recommended Baidu optimizations that rearrange the order of computations are probably needed to get the effect.

**Comment by nouiz** [comment]
On the GPU, there could be small speed up with your numbers. But it is
small ~6%. It could be noise too.

**Comment by carlthome** [comment]
@nouiz, @gw0, maybe let the experiment run a little longer (a couple of minutes per setting, like 20 epochs or so) just to make sure?

Also, how many timesteps is BPTT truncated to?

**Comment by gw0** [comment]
@carlthome Maybe, but it would make much more sense to implement all Baidu recommendations and then do a thorough evaluation. Unfortunately I do not have time for this, can you implement it?

The benchmark was based on https://github.com/fchollet/keras/blob/master/examples/lstm_benchmark.py, therefore 80 timesteps are used and backpropagation though time is not truncated.

**Comment by carlthome** [comment]
@gw0, sorry, unfortunately I don't have the time.
